docs/javascript/render-functions.md

```diff
@@ -0,0 +1,292 @@
+# React on Rails Render Functions: Usage Guide
+
+Based on the test file `serverRenderReactComponent.test.ts` and the existing documentation, I'll clarify how render functions work in React on Rails and how to use them with each Ruby helper method.
+
+## Types of Render Functions and Their Return Values
+
+Looking at the test file, render functions can return several types of values:
+
+### 1. React Components (JSX)
+
+```jsx
+const MyComponent = (props, _railsContext) => {
+  return <div>Hello {props.name}</div>;
+};
+```
+
+### 2. Objects with renderedHtml Property
+
+```jsx
+const MyComponent = (props, _railsContext) => {
+  return {
+    renderedHtml: `<div>Hello ${props.name}</div>`,
+  };
+};
+```
+
+### 3. Objects with Multiple HTML Properties
+
+```jsx
+const MyComponent = (props, _railsContext) => {
+  return {
+    componentHtml: <div>Hello {props.name}</div>,
+    title: `<title>${props.title}</title>`,
+    metaTags: `<meta name="description" content="${props.description}" />`,
+  };
+};
+```
+
+### 4. Promises of Strings
+
+```jsx
+const MyComponent = async (props, _railsContext) => {
+  const data = await fetchData();
+  return `<div>Hello ${data.name}</div>`;
+};
+```
+
+### 5. Promises of Objects
+
+```jsx
+const MyComponent = async (props, _railsContext) => {
+  const data = await fetchData();
+  return {
+    componentHtml: `<div>Hello ${data.name}</div>`,
+  };
+};
+```
+
+### 6. Redirect Information
+
+```jsx
+const MyComponent = (props, _railsContext) => {
+  return {
+    redirectLocation: '/new-path',
+    error: null,
+    renderedHtml: null,
+  };
+};
+```
+
+## Important Limitations Observed in Tests
+
+The test file reveals some important limitations:
+
+1. **Direct String Returns Don't Work** - Returning a raw HTML string directly from a render function causes an error. The test `doesn't render html string returned directly from render function` demonstrates this.
+
+2. **Non-Promise Objects Need renderedHtml** - Non-promise objects must include a `renderedHtml` property to be valid, as shown in the test `doesn't render object without renderedHtml property`.
+
+3. **Async Functions Have Different Behavior** - Interestingly, the test `returns the object returned by async render function even if it doesn't have renderedHtml property` shows that Promise-returning functions can return objects without a `renderedHtml` property and they will still work.
```

Comment:
Retitle this section to provide greater clarity, remove "interestly" and maybe add the reason why async functions do not require the `renderedHtml` property.